src/compiler/parser.ts

```diff
@@ -1811,7 +1811,7 @@ namespace ts {
         function parseEntityName(allowReservedWords: boolean, diagnosticMessage?: DiagnosticMessage): EntityName {
             let entity: EntityName = parseIdentifier(diagnosticMessage);
             while (parseOptional(SyntaxKind.DotToken)) {
-                const node = <QualifiedName>createNode(SyntaxKind.QualifiedName, entity.pos);
+                const node: QualifiedName = <QualifiedName>createNode(SyntaxKind.QualifiedName, entity.pos);  // !!!

```

Comment:
Just to make the issue into a concrete, self-contained test, I'll drop this here:
```ts
interface First {first: 'first'}
interface Second {second: 'second'}
interface Third {third: 'third'}

declare function next(obj: Third): undefined;
declare function next(obj: Second): Third;
declare function next(obj: First): Second;

declare function isThird(x: any): x is Third;

var x: First | Second | Third;
x = {first: 'first'};
x.first;
while (true) {
    x = next(x);
    x.second; // Should be forbidden (x is Third in the second iteration), but is allowed
    if (isThird(x)) {
        break;
    }
}

x.third; // Should be allowed (loop always exits when x is Third), but is forbidden
```
Even if we don't _fix_ the issue, the current behavior can be awkward. (`x` ends up typed as `Second` after the loop, which is incorrect.)